doc/design/backfill-starvation-prevention.md

```diff
@@ -0,0 +1,179 @@
+
+type TaskInfo struct {
+  UID         TaskID
+  Job         JobID
+  Name        string
+  Namespace   string
+  Resreq      *Resource
+  InitResreq  *Resource
+  NodeName    string
+  Status      TaskStatus
+  Priority    int32
+  VolumeReady bool
+  Pod         *v1.Pod
+  IsBackfill  bool
+}
+
+func (ji *JobInfo) GetReadiness() JobReadiness {
+  if allocatedTasksCnt >= ji.MinAvailable {
+     return Ready
+  } else if conditionallyAllocatedTasksCnt >= ji.MinAvailable {
+     return ConditionallyReady
+  }
+  return NotReady
+}
+```
+
+### NodeInfo
+* Add `Backfilled` Resource in NodeInfo. Backfilled resources are resources occupied by backfill tasks.
+* Add `GetAccessibleResource` method in NodeInfo. Accessible resource is the sum of idle resource and backfilled resource. 
+
+```golang
+type NodeInfo struct {
+  Name string
+  Node *v1.Node
+  Releasing *Resource
+  Idle *Resource
+  Used *Resource
+  Allocatable *Resource
+  Capability  *Resource
+  Tasks map[TaskID]*TaskInfo
+  Backfilled *Resource
+}
+
+func (ni *NodeInfo) GetAccessibleResource() *Resource {
+  idle := ni.Idle.Clone()
+  return idle.Add(ni.Backfilled)
+}
+```
+	 
+The following diagram illustrated the relations between *Allocatable*, *Idle*, *Used*, *Accessible*, and *Backfilled* in `NodeInfo`. Accessible resources of a node is the sum of the node's idle resources and backfilled resources. Task T1 is an allocatable task because its request amount is less than the node’s idle amount. Task T2 is a conditionally allocatable task because its request amount is greater than the node’s idle amount, but less than the node’s accessible amount. Task T3 is an unallocable task because its request amount is greater than the node’s accessible amount.
+
+![node-resource](../images/node-resource.png)
+ 
+## Algorithm
+
+Kube batch scheduler does scheduling in rounds. In each round, it runs the following four scheduling actions in order: reclaim, allocate, preempt, backfill. The actual action sequence is configurable. 
+
+This section describes the gist of the algorithm. To visualize how the algorithm works, please take a look at [the slides](https://docs.google.com/presentation/d/1kWbR0_lRGSjkgCnUoySNUktIqKVcR2OOTsyq7WtUIGI/edit#slide=id.g509241a485_0_1041).
+
+### Allocation Phase
+* In allocation phase, we allocate resources on a node to a task, if either of the following two conditions is true
+  * The job is starvating. If a job is starving, we will allocate the remaining resources (of a node) to the task even if the node’s remaining resource is less than the task’s request resource. By doing so, we prevent the scheduler from allocating the remaining resources to other jobs. In other words, we reserve the remaining resources for the starving job. Eventually, some running jobs on the node will finish running, and the starving task will get enough resource to start. This is how we prevent a job from be starved for ever.
```

Comment:
How to choose which node a task will reserve? How many nodes will it reserve?

Reply:
Node ordering algorithm determines which node to reserve. One node is reserved for each task.

Reply:
I am afraid that it might be not enough to reserve just one node for some cases, e.g. tasks on that node might run for a long time. Will the task always wait on the same node for resource?